src/shrinkage.jl

```diff
@@ -1,91 +1,46 @@
-"""
-function shrinkage(m::LinearMixedModel{T}) where {T}
-    # should this be fixefnames or coefnames?
-    # need to think about when pivoting is occuring
-    fenms = fixefnames(m)
-    # returns a view from Xymat in MM4, or just the response vec in MM3.x
-    yvec = response(m)
-    ranefs = ranef(m)
-    cvec = sizehint!(CoefByGroup[], length(ranefs))
-    for (j, re) in enumerate(m.reterms)
-        cnms = re.cnames
-        cnms ⊆ fenms || throw(ArgumentError("No corresponding fixed effect for random effects $(setdiff(cnms, fenms)): there is no estimated grand mean to measure shrinkage towards"))
-        # XXX Should m.X return a view into m.Xymat?
-        Xmat = view(m.X, :, [findfirst(==(nm), fenms) for nm in cnms])
-        levs = re.levels
-        refs = re.refs
-        grpest = Matrix{Union{Missing,T}}(missing, (length(levs), length(cnms)))
-        for i in eachindex(levs)
-            rows = findall(==(i), refs)
-            try
-                grpest[i, :] = view(Xmat, rows, :) \ view(yvec, rows)
-            catch
-            end
-        end
-        push!(cvec, CoefByGroup(cnms, levs, Xmat \ yvec, ranefs[j]', grpest))
+function shrinkageplot(
+    m::LinearMixedModel{T},
+    gf::Symbol=first(fnames(m)),
+    θref::AbstractVector{T}=10000 .* m.optsum.initial,
+) where {T}
+    reind = findfirst(==(gf), fnames(m))  # convert the symbol gf to an index
+    if isnothing(reind)
+        throw(ArgumentError("gf=$gf is not one of the grouping factor names, $(fnames(m))"))
     end
-    NamedTuple{fnames(m)}((cvec...,))
-end
-
-shrinkageplot(m::LinearMixedModel, gf::Symbol=first(fnames(m))) = shrinkageplot(shrinkage(m)[gf])
-
-shrinkageplot(cg::CoefByGroup) = shrinkageplot!(Figure(resolution=(1000,1000)), cg)
-
-function shrinkageplot!(f::Figure, cg::CoefByGroup)
-    shrinkage2d!(Axis(f[1,1]), cg)
-    f
-end
-
-function shrinkage2d!(a::Axis, cg::CoefByGroup{T}, inds=(1, 2)) where T
-    i, j = inds
-    fxd = cg.fixed
-    scatter!(a, view(fxd, i:i), view(fxd, j:j), color=:green, label="Pop")
-    u = view(cg.condmodes, :, i) .+ fxd[i]
-    v = view(cg.condmodes, :, j) .+ fxd[j]
-    scatter!(a, u, v, color=:blue, label="Mixed")
-    x = view(cg.grpest, :, i)
-    y = view(cg.grpest, :, j)
-    missgrp = ismissing.(x) .| ismissing.(y)
-    if any(missgrp)
-        nonmiss = .!(missgrp)
-        x = convert(Vector{T}, view(x, nonmiss))
-        y = convert(Vector{T}, view(y, nonmiss))
-        u = view(u, nonmiss)
-        v = view(v, nonmiss)
-    else
-        x = convert(Vector{T}, x)
-        y = convert(Vector{T}, y)
+    reest = ranef(m)[reind]               # random effects conditional means at estimated θ
+    reref = ranef(updateL!(setθ!(m, θref)))[reind]  # same at θref
+    updateL!(setθ!(m, m.optsum.final))    # restore parameter estimates and update m
+    cnms = m.reterms[reind].cnames
+    f = Figure(; resolution=(1000, 1000)) # use an aspect ratio of 1 for the whole figure
```

Comment:
Perhaps.  As I mention above, it would be best to get a feeling for how we want to design the system first.  The nice thing about the Figure is that it has accessors for the information on the layout, etc. Of course, we could always use `shrinkageplot!` to modify a Figure and have `shrinkageplot` just create a default Figure and pass it to `shrinkageplot!`.

Reply:
Right, right. I'm happy to be in a high-churn period right now while we figure out what a good interface would look like. (And in that regard, we're no worse than the rest of the Makie ecoystem :smile: )